contracts/nft/erc721m/ERC721CMInitializableRedeemable.sol

```diff
@@ -0,0 +1,636 @@
+        return _contractURI;
+    }
+
+    /*==============================================================
+    =                             MODIFIERS                        =
+    ==============================================================*/
+
+    /// @notice Modifier to check if the contract is mintable
+    modifier canMint() {
+        if (!_mintable) revert NotMintable();
+        _;
+    }
+
+    /// @notice Modifier to check if the total supply is enough
+    /// @param qty The quantity to mint
+    modifier hasSupply(uint256 qty) {
+        if (totalSupply() + qty > _maxMintableSupply) revert NoSupplyLeft();
+        _;
+    }
+
+    /*==============================================================
+    =                     PUBLIC WRITE METHODS                     =
+    ==============================================================*/
+
+    /// @notice Mints tokens for the caller
+    /// @param qty The quantity to mint
+    /// @param limit The minting limit for the caller (used in merkle proofs)
+    /// @param proof The merkle proof for allowlist minting
+    /// @param timestamp The timestamp for the minting action (used in cosigning)
+    /// @param signature The cosigner's signature
+    function mint(uint32 qty, uint32 limit, bytes32[] calldata proof, uint256 timestamp, bytes calldata signature)
+        external
+        payable
+        virtual
+        nonReentrant
+    {
+        _mintInternal(qty, msg.sender, limit, proof, timestamp, signature);
+    }
+
+    /// @notice Allows authorized minters to mint tokens for a specified address
+    /// @param to The address to mint tokens for
+    /// @param qty The quantity to mint
+    /// @param limit The minting limit for the recipient (used in merkle proofs)
+    /// @param proof The merkle proof for allowlist minting
+    /// @param timestamp The timestamp for the minting action (used in cosigning)
+    /// @param signature The cosigner's signature
+    function authorizedMint(
+        uint32 qty,
+        address to,
+        uint32 limit,
+        bytes32[] calldata proof,
+        uint256 timestamp,
+        bytes calldata signature
+    ) external payable onlyAuthorizedMinter {
+        _mintInternal(qty, to, limit, proof, timestamp, signature);
+    }
+
+    /// @notice Burns a specific token.
+    /// @dev Only callable by the token owner or an approved operator. The token must exist.
+    /// @param tokenId The ID of the token to burn.
+    function burn(uint256 tokenId) external {
+        _burn(tokenId, true);
+    }
+
+    /// @notice Burns a list of tokens.
+    /// @dev Only callable by the REDEEMER_ROLE, no approval required. The token must exist.
+    /// @param from The owners of the tokens to burn.
+    /// @param tokenIds The IDs of the tokens to burn.
+    function redeem(address[] memory from, uint256[] memory tokenIds) external onlyAuthorizedRedeemer {
+        for (uint256 i = 0; i < tokenIds.length; i++) {
```

Comment:
Please add a check for from.length === tokenIds.length

Reply:
Can we get rid of `from` and use `ownerOf` to get owner address?

Reply:
see: https://github.com/me-foundation/magicdrop/pull/180#discussion_r1970954615